x/xgit/repository.go

```diff
@@ -3,11 +3,44 @@ package xgit
 import (
 	"errors"
 	"net/url"
+	"strings"
 
 	git "gopkg.in/src-d/go-git.v4"
 	"gopkg.in/src-d/go-git.v4/plumbing"
+	"gopkg.in/src-d/go-git.v4/plumbing/protocol/packp"
+	"gopkg.in/src-d/go-git.v4/plumbing/transport"
+	"gopkg.in/src-d/go-git.v4/plumbing/transport/client"
 )
 
+const defaultScheme = "https"
+
+// IsGitURL checks if given URL is git repository.
+func IsGitURL(URL string) bool {
+	if strings.HasPrefix(URL, "git://") || strings.HasSuffix(URL, ".git") {
+		return true
+	}
+	_, err := ListRemote(URL)
+	return err == nil
+}
+
+// ListRemote lists all remotes from given url.
+func ListRemote(URL string) (*packp.AdvRefs, error) {
+	URL = addDefaultURLScheme(URL)
+	ep, err := transport.NewEndpoint(URL)
+	if err != nil {
+		return nil, err
+	}
+	c, err := client.NewClient(ep)
+	if err != nil {
+		return nil, err
+	}
+	s, err := c.NewUploadPackSession(ep, nil)
+	if err != nil {
+		return nil, err
+	}
+	return s.AdvertisedReferences()
```

Comment:
Is it the only way to test if the url is a git repository?

Reply:
Is there some more generic version? Like with `content-accept` header?

Reply:
It will be better if the system and generic and we could be compatible with lots of sources by just implementing a interface.

Reply:
Is it what the code is doing?

```
git ls-remote --exit-code -h "$REPO_URL"
```